File: src/test/java/com/wave/hextractor/object/HexTableTest.java
package com.wave.hextractor.object;

import com.wave.hextractor.pojo.OffsetEntry;
import org.junit.Assert;
import org.junit.Test;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

public class HexTableTest {

    private static final List<String> LINES = Collections.unmodifiableList(Arrays.asList(
            "00=a", "01=b", "02=c", "03=ab" ,"ad"
    ));

    private static final String LINESRES = "00=á\n" +
            "00=a\n" +
            "03=ab\n" +
            "01=b\n" +
            "02=c\n";

    private static final List<String> LINES2 = Collections.unmodifiableList(Arrays.asList(
            "00=a", "01=A", "02=c", "03=ab" ,"ad"
    ));

    private static final List<String> LINES3 = Collections.unmodifiableList(Arrays.asList(
            "00=a", "01=A", "02=0", "03=ab" ,"ad"
    ));

    @Test
    public void constructors() {
        boolean ok;
        try {
            HexTable test = new HexTable();
            Assert.assertNotNull(test.toString());
            test = new HexTable(LINES);
            Assert.assertNotNull(test.toString());
            ok = true;
        }
        catch(Exception e) {
            ok = false;
        }
        Assert.assertTrue(ok);
    }

    @Test
    public void toString1() {
        HexTable table = new HexTable(LINES);
        Assert.assertEquals("b", table.toString((byte)1, true, true));
        Assert.assertEquals(".", table.toString((byte)3, false, false));
        Assert.assertEquals("ab", table.toString((byte)3, true, false));
        Assert.assertEquals("~04~", table.toString((byte)4, false, true));
    }

    @Test
    public void toString2() {
        HexTable table = new HexTable(LINES);
        Assert.assertEquals(".", table.toString((byte)3, false));
        Assert.assertEquals("ab", table.toString((byte)3, true));
    }

    @Test
    public void toSelectionString() {
        HexTable table = new HexTable(LINES);
        HexTable table2 = new HexTable(LINES2);
        HexTable table3 = new HexTable(LINES3);
        Assert.assertEquals("a: 00 ", table.toSelectionString());
        Assert.assertEquals("A: 01 a: 00 ", table2.toSelectionString());
        Assert.assertEquals("A: 01 a: 00 0: 02 ", table3.toSelectionString());
    }

    @Test
    public void toAscii() {
        HexTable table = new HexTable(LINES);
        Assert.assertEquals("abc..", table.toAscii(new byte[] {0,1,2,3,4}, false, false));
        Assert.assertEquals("abcab.", table.toAscii(new byte[] {0,1,2,3,4}, true, false));
        Assert.assertEquals("abcab~04~", table.toAscii(new byte[] {0,1,2,3,4}, true, true));
        Assert.assertEquals("abc~03~~04~", table.toAscii(new byte[] {0,1,2,3,4}, false, true));
    }

    @Test
    public void toAscii1() {
        HexTable table = new HexTable(LINES);
        Assert.assertEquals("abc..", table.toAscii(new byte[]{0, 1, 2, 3, 4}, false));
        Assert.assertEquals("abcab.", table.toAscii(new byte[]{0, 1, 2, 3, 4}, true));
    }

    @Test
    public void toHex() {
        HexTable table = new HexTable(LINES);
        byte[] res = {0, 1, 2, 0, 0, 1, 0};
        Assert.assertArrayEquals(res, table.toHex("abc{ab}"));
    }

    @Test
    public void addToTable() {
        HexTable table = new HexTable();
        byte[] data = {5, 0};
        Assert.assertNotEquals("Ka", table.toAscii(data, false));
        table.addToTable((byte) 5, "K");
        table.addToTable((byte) 0, "a");
        Assert.assertEquals("Ka", table.toAscii(data, false));
    }

    @Test
    public void toAscii2() {
        HexTable table = new HexTable(LINES);
        byte[] data = new byte[]{0, 1, 2, 3, 4};
        OffsetEntry entry = new OffsetEntry(0, 4, Arrays.asList("FF"));
        String ascii = "@00000000-00000004-FF\n" +
                ";00000000{abc{ab}~04~}#011#005\n" +
                "abc{ab}~04~#005\n" +
                "|5\n";
        Assert.assertEquals(ascii, table.toAscii(data, entry,true));
        Assert.assertEquals(ascii, table.toAscii(data, entry,false));
    }

    @Test
    public void toHex1() {
        HexTable table = new HexTable(LINES);
        byte[] data = new byte[]{0, 1, 2, 3, 4};
        OffsetEntry entry = new OffsetEntry(0, 4, Arrays.asList("FF"));
        Assert.assertArrayEquals(data, table.toHex("abc{ab}~04~#005\n|5\n", entry));
    }

    @Test
    public void getAllEntries() throws IOException {
        HexTable table = new HexTable(0);
        File searchAll = new File(Objects.requireNonNull(getClass().getClassLoader().getResource("files/searchAll.txt")).getFile());
        byte[] secondFileBytes = Files.readAllBytes(searchAll.toPath());
        File dictFile = new File(Objects.requireNonNull(getClass().getClassLoader().getResource("files/TestDict.txt")).getFile());
        Assert.assertEquals("00000000-0000000A-00-FF,0000000C-00000015-00-FF,00000022-0000002B-00-FF,",
                table.getAllEntries(secondFileBytes, 4, 1,
        Arrays.asList("00", "FF"), dictFile.getAbsolutePath()));
    }

    @Test
    public void toAsciiTable() {
        HexTable table = new HexTable(LINES);
        Assert.assertEquals(LINESRES, table.toAsciiTable());
    }

    @Test
    public void hashCode1() {
        HexTable table1 = new HexTable();
        HexTable table2 = new HexTable();
        Assert.assertEquals(table1.hashCode(), table2.hashCode());
        HexTable table3 = new HexTable(LINES);
        HexTable table4 = new HexTable(LINES);
        Assert.assertEquals(table3.hashCode(), table4.hashCode());
    }

    @Test
    public void equals1() {
        HexTable table1 = new HexTable();
        HexTable table2 = new HexTable();
        Assert.assertEquals(table1, table2);
        HexTable table3 = new HexTable(LINES);
        HexTable table4 = new HexTable(LINES);
        Assert.assertEquals(table3, table4);
        Assert.assertNotEquals(table3, new Object());
    }

    @Test
    public void getSearchPercent() {
        HexTable table = new HexTable(LINES);
        Assert.assertTrue(table.getSearchPercent() == 0);
    }
}

File: src/test/kotlin/com/wave/hextractor/helper/HexTableTest.kt
package com.wave.hextractor.helper

import com.wave.hextractor.pojo.OffsetEntry
import org.junit.Assert
import org.junit.Test
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.util.*

class HexTableTest {
    @Test
    fun constructors() {
        var ok: Boolean
        try {
            var test = HexTable()
            Assert.assertNotNull(test.toString())
            test = HexTable(LINES)
            Assert.assertNotNull(test.toString())
            ok = true
        } catch (e: Exception) {
            ok = false
        }
        Assert.assertTrue(ok)
    }

    @Test
    fun toString1() {
        val table = HexTable(LINES)
        Assert.assertEquals("b", table.toString(1.toByte(), expand = true, decodeUnknown = true))
        Assert.assertEquals(".", table.toString(3.toByte(), expand = false, decodeUnknown = false))
        Assert.assertEquals("ab", table.toString(3.toByte(), expand = true, decodeUnknown = false))
        Assert.assertEquals("~04~", table.toString(4.toByte(), expand = false, decodeUnknown = true))
    }

    @Test
    fun toString2() {
        val table = HexTable(LINES)
        Assert.assertEquals(".", table.toString(3.toByte(), false))
        Assert.assertEquals("ab", table.toString(3.toByte(), true))
    }

    @Test
    fun toSelectionString() {
        val table = HexTable(LINES)
        val table2 = HexTable(LINES2)
        val table3 = HexTable(LINES3)
        Assert.assertEquals("a: 00 ", table.toSelectionString())
        Assert.assertEquals("A: 01 a: 00 ", table2.toSelectionString())
        Assert.assertEquals("A: 01 a: 00 0: 02 ", table3.toSelectionString())
    }

    @Test
    fun toAscii() {
        val table = HexTable(LINES)
        Assert.assertEquals("abc..", table.toAscii(byteArrayOf(0, 1, 2, 3, 4), expand = false, false))
        Assert.assertEquals("abcab.", table.toAscii(byteArrayOf(0, 1, 2, 3, 4), expand = true, decodeUnknown = false))
        Assert.assertEquals("abcab~04~", table.toAscii(byteArrayOf(0, 1, 2, 3, 4), expand = true, decodeUnknown = true))
        Assert.assertEquals("abc~03~~04~", table.toAscii(byteArrayOf(0, 1, 2, 3, 4),
            expand = false,
            decodeUnknown = true
        ))
    }

    @Test
    fun toAscii1() {
        val table = HexTable(LINES)
        Assert.assertEquals("abc..", table.toAscii(byteArrayOf(0, 1, 2, 3, 4), false))
        Assert.assertEquals("abcab.", table.toAscii(byteArrayOf(0, 1, 2, 3, 4), true))
    }

    @Test
    fun toHex() {
        val table = HexTable(LINES)
        val res = byteArrayOf(0, 1, 2, 0, 0, 1, 0)
        Assert.assertArrayEquals(res, table.toHex("abc{ab}"))
    }

    @Test
    fun addToTable() {
        val table = HexTable()
        val data = byteArrayOf(5, 0)
        Assert.assertNotEquals("Ka", table.toAscii(data, false))
        table.addToTable(5.toByte(), "K")
        table.addToTable(0.toByte(), "a")
        Assert.assertEquals("Ka", table.toAscii(data, false))
    }

    @Test
    fun toAscii2() {
        val table = HexTable(LINES)
        val data = byteArrayOf(0, 1, 2, 3, 4)
        val entry = OffsetEntry(0, 4, mutableListOf("FF"))
        val ascii = """
            @00000000-00000004-FF
            ;00000000{abc{ab}~04~}#011#005
            abc{ab}~04~#005
            |5
            
        """.trimIndent()
        Assert.assertEquals(ascii, table.toAscii(data, entry, true))
        Assert.assertEquals(ascii, table.toAscii(data, entry, false))
    }

    @Test
    fun toHex1() {
        val table = HexTable(LINES)
        val data = byteArrayOf(0, 1, 2, 3, 4)
        val entry = OffsetEntry(0, 4, mutableListOf("FF"))
        Assert.assertArrayEquals(data, table.toHex("abc{ab}~04~#005\n|5\n", entry))
    }

    @Throws(IOException::class)
    @Test
    fun allEntries() {
        val table = HexTable(0)
        val searchAll = File(Objects.requireNonNull(javaClass.classLoader.getResource("files/searchAll.txt")).file)
        val secondFileBytes = Files.readAllBytes(searchAll.toPath())
        val dictFile = File(Objects.requireNonNull(javaClass.classLoader.getResource("files/TestDict.txt")).file)
        Assert.assertEquals(
            "00000000-0000000A-00-FF,0000000C-00000015-00-FF,00000022-0000002B-00-FF,",
            table.getAllEntries(secondFileBytes, 4, 1, mutableListOf("00", "FF"), dictFile.absolutePath)
        )
    }

    @Test
    fun toAsciiTable() {
        val table = HexTable(LINES)
        Assert.assertEquals(LINESRES, table.toAsciiTable())
    }

    @Test
    fun hashCode1() {
        val table1 = HexTable()
        val table2 = HexTable()
        Assert.assertEquals(table1.hashCode().toLong(), table2.hashCode().toLong())
        val table3 = HexTable(LINES)
        val table4 = HexTable(LINES)
        Assert.assertEquals(table3.hashCode().toLong(), table4.hashCode().toLong())
    }

    @Test
    fun equals1() {
        val table1 = HexTable()
        val table2 = HexTable()
        Assert.assertEquals(table1, table2)
        val table3 = HexTable(LINES)
        val table4 = HexTable(LINES)
        Assert.assertEquals(table3, table4)
        Assert.assertNotEquals(table3, Any())
    }

    @Test
    fun searchPercent() {
        val table = HexTable(LINES)
        Assert.assertEquals(0.0, table.searchPercent.toDouble(), 0.0)
    }

    companion object {
        private val LINES = Collections.unmodifiableList(
            mutableListOf(
                "00=a",
                "01=b",
                "02=c",
                "03=ab",
                "ad"
            )
        )
        private const val LINESRES = "00=á\n" + "00=a\n" + "03=ab\n" + "01=b\n" + "02=c\n"
        private val LINES2 = Collections.unmodifiableList(
            mutableListOf(
                "00=a",
                "01=A",
                "02=c",
                "03=ab",
                "ad"
            )
        )
        private val LINES3 = Collections.unmodifiableList(
            mutableListOf(
                "00=a",
                "01=A",
                "02=0",
                "03=ab",
                "ad"
            )
        )
    }
}
